server/base/src/main/java/org/apache/accumulo/server/util/SystemPropUtil.java

```diff
@@ -70,6 +81,20 @@ public static void setSystemProperty(ServerContext context, String property, Str
 
   public static void removeSystemProperty(ServerContext context, String property)
       throws InterruptedException, KeeperException {
+    AtomicBoolean shouldRemove = new AtomicBoolean(true);
+    DeprecatedPropertyUtil.getReplacementName(property, (log, replacement) -> {
+      log.warn("{} was deprecated and will be removed in a future release;"
+          + " no action was taken because it is not set here;"
+          + " did you mean to remove its replacment {} instead?", property, replacement);
+      shouldRemove.set(false);
+    });
+    if (shouldRemove.get()) {
+      removePropWithoutDeprecationWarning(context, property);
+    }
```

Comment:
@keith-turner I'm wondering if you have any opinion on this change in particular. This change means a user would not be able to use the old property name when removing a property from zookeeper--they could only do that using the new property name. You had mentioned some concern about automation updating zookeeper properties, though setting a property using the old name is supported. Do you have any concerns here? The thinking is for a property removal it is more dangerous to assume the user meant to remove the new property name and used the old name.

Reply:
What is the rationale for this asymmetrical behavior?  Set and delete both mutate a property.  One could think of delete as special case of set that sets the value to NULL.

Reply:
There's a discussion @ctubbsii and I had [here](https://github.com/brianloss/accumulo/pull/2#discussion_r562689513).

Reply:
@keith-turner The rationale is that when setting a deprecated property, the intent is clear that the user is trying to change the behavior of Accumulo by setting the property to some specific value. We can handle that intent by translating to the new name when we store it. However, when a user is removing a property, the intent is much less clear, and could be *very* unintuitive. It would be very strange, for example, to ask to remove "X" (which does not exist), and then to have "Y" deleted instead. Rather than rely on this unintuitive behavior, I think it would be better to just do the safest thing and make no changes, relying instead on the very verbose warning, so that the user can be more explicit about what they intended to do.

Furthermore, removing a property already had the asymmetrical behavior of not triggering any warning or error when removing a property that isn't a valid zoo property, while setting the same property would have triggered an error. The behavior in this PR is an extension of that existing NOOP behavior when an invalid property is requested to be removed. This preserves the previous behavior of the remove method, only adding a new warning.

Honestly, I preferred the solution where we just prevent the API from being able to set the old `master.*` property names entirely. That way, it would just get an error, and we wouldn't have to try to discern and accommodate user intent. Yes, in the edge case where there might be some automation setting/removing properties, that automation may fail as a result of no longer being able to set/remove these properties, since they aren't recognized. But, I think that would be far more preferable than trying to guess user intent and try to communicate with users via warnings, introducing the possibility for more mistakes (mistakes due to not expecting the automatic translation, or mistakes due to expecting an automatic translation that doesn't occur). At least with an error, the situation is clearly communicated to the caller.

Reply:
I think a nice transitional experience is that each manager property has a deprecated alias master property that can be used interchangeably(with a warning).  This is being achieved for set, but not delete and I think the semantics are very confusing. 

> Honestly, I preferred the solution where we just prevent the API from being able to set the old master.* property names entirely. 

When I put myself in the shoes of a user, that is not the experience I would want.  An experience where code that worked in 2.0 (code written following Accumulo documentation) blows up by design in 2.1 (and its only detectable at runtime).  As a user I would prefer the experience where it continues to work in 2.1 (with warnings) and blows up at runtime in 2.2, 3.0, etc.

If master aliases are only supported for set and not delete, I think it may be a less confusing user experience to not support aliases for either set or delete and throw errors for both instead.  Errors are not my personal preference, but its consistent at least and not confusing.

Reply:
> As a user I would prefer the experience where it continues to work in 2.1 (with warnings) and blows up at runtime in 2.2, 3.0, etc.

"it continues to work" is something I would prefer also, but it's not that simple, because there isn't a single behavior that is expected. There are multiple distinct behaviors, and the phrase "it continues to work" is ambiguous about which ones should be preserved and which ones are acceptable to be changed.

Here are some of the behaviors of 2.0 that people might expect:

A. Configurability: `{ } -> set X -> { X }`
B. Independence: `{ X } -> set Y -> { X, Y } -> remove X -> { Y }`
C. Reversibility: `{ } -> set X -> { X } -> remove X -> { }`
D. NOOP on removing missing: `{ Y } -> remove X -> { Y }`

With the automatic translation of of `X -> Y`, no matter what, some of these expectations are going to be broken with any of the proposed behaviors for 2.1:

1. `{ } -> set X -> { Y } (with WARN)` - breaks independence (the intent of this PR, so this break is acceptable when limited to the properties being upgraded)
2. `{ Y } -> remove X -> { } (with WARN)` - your proposal; preserves reversibility, but breaks NOOP on removing missing
3. `{ Y } -> remove X -> { Y } (with WARN)` - my proposal; preserves NOOP on removing missing, but preserves reversibility
4. `{ Y } -> remove X -> ERROR` - breaks NOOP on removing missing
5. `{ Y } -> set X -> ERROR` - breaks configurability of the old properties, but at least could be consistent with option 4 if used in conjunction with that

> Errors are not my personal preference, but its consistent at least and not confusing.

If we accept that there are different user expectations and we can't reliably infer user intent, then at least we can be unambiguous and perfectly clear. However, I'm not certain if it makes more sense to be clear with only ERROR messages, or also with thrown exceptions, because adding new thrown exceptions on the remove case is breaking existing behavior (NOOP on removing missing). For the set case, it wouldn't be that unusual, since we already have some properties which cannot be set in ZooKeeper, which generate an exception. These same properties do not generate a corresponding exception when requested to be removed from ZooKeeper, so there's already an asymmetry there.

Reply:
I'm in the camp of removing the replacement property if the deprecated property name is sent to remove. I've been looking at this change though we're telling users: "Property X has been replaced by property Y, but in order to be less disruptive we are going to continue (for a short time) to allow you to use X and behind the scenes we'll replace it with Y. If you use X, it's as though you had used Y instead."

Given that we will have documented that using X is a substitute for using Y, I'm struggling to come up with a good example where it is confusing or unexpected for the user if they use the deprecated property name with remove and we remove the replacement property (I'd say it's confusing not to do it). The only case I can come with is if the user chooses to immediately start using the new property names and then thinks they need to clean up zookeeper themselves and remove the old properties so they call remove on the old names. That usage is inconsistent with the documentation, and they'd figure it out quickly with the warning that we're removing the replacement property names instead. Are there other ways it could be confusing/unexpected for the user if we remove the replacement property when they attempt to remove the deprecated property?

Reply:
Okay. That's fine. I'm not entirely convinced, but I've made my case, and I am in the minority. I can submit another update to Brian's code to apply the behavior advocated for by Keith and Brian. I appreciate both of your patience in letting me advocate for my perspective, but I'm content to let it go at this point, and proceed with the majority viewpoint.

Reply:
@keith-turner and @brianloss - I pushed the changes based on this discussion directly to this PR in fa0db45e8c59ec566c5f9f25582bd01ae234f8bc